File: com/smartadserver/android/library/mediation/SASGMACustomEventUtil.java
package com.smartadserver.android.library.mediation;

import android.content.Context;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.gms.ads.mediation.MediationAdRequest;
import com.google.android.gms.ads.mediation.VersionInfo;
import com.smartadserver.android.library.model.SASAdPlacement;
import com.smartadserver.android.library.util.SASConfiguration;
import com.smartadserver.android.library.util.SASLibraryInfo;

import java.util.Set;

/**
 * Base class of all other SASCustomEvent adapters. Handle the SDK configuration.
 */
public class SASGMACustomEventUtil {

    public final static String MEDIATION_EXTRAS_SMART_KEYWORD_TARGETING_KEY = "smart_keyword_targeting";

    // version info needed by google classes, init in a static block
    private static VersionInfo adapterVersionInfo = null;
    private static VersionInfo sdkVersionInfo = null;

    static {
        adapterVersionInfo = new VersionInfo(1,0,0);

        int majorVersion = 0;
        int minorVersion = 0;
        int microVersion = 0;

        String[] versionInfo = SASLibraryInfo.getSharedInstance().getVersion().split(".");

        // we expect 3 tokens, no more, no less
        if (versionInfo.length >= 1) {
            try {
                majorVersion = Integer.parseInt(versionInfo[0]);
            } catch (NumberFormatException ignored) {}
        }
        if (versionInfo.length >= 2) {
            try {
                minorVersion = Integer.parseInt(versionInfo[1]);
            } catch (NumberFormatException ignored) {}
        }
        if (versionInfo.length >= 3) {
            try {
                microVersion = Integer.parseInt(versionInfo[2]);
            } catch (NumberFormatException ignored) {}
        }

        sdkVersionInfo = new VersionInfo(majorVersion, minorVersion, microVersion);
    }

    /**
     * Configure the Smart Display SDK if needed and
     *
     * @param context         The application context.
     * @param placementString The Smart placement String.
     * @param mediationExtras a Bundle containing mediation extra parameters as passed to the Google AdRequest in the application
     * @return a valid SASAdPlacement, or null if the SDK can not be configured or if the placement string is wrongly set.
     */
    @Nullable
    public static SASAdPlacement configureSDKAndGetAdPlacement(@NonNull Context context,
                                                               @NonNull String placementString,
                                                               @Nullable Bundle mediationExtras) {

        int siteId = -1;
        String pageId = "";
        int formatId = -1;
        String targeting = "";

        // tokenize placement string and fill adPlacement;
        String[] ids = placementString.split("/");
        if (ids.length >= 3) {
            try {
                siteId = Integer.parseInt(ids[0].trim());
                pageId = ids[1].trim();
                formatId = Integer.parseInt(ids[2].trim());
            } catch (Exception e) {
                // invalid placement, return null
                return null;
            }
        }

        // configure the Smart Ad Server SDK if necessary
        if (!SASConfiguration.getSharedInstance().isConfigured()) {
            try {
                if (siteId >= 1) {
                    SASConfiguration.getSharedInstance().configure(context, siteId);
                } else {
                    return null;
                }
            } catch (SASConfiguration.ConfigurationException e) {
                e.printStackTrace();
                return null;
            }
        }

        // extract custom targeting from mediation extras
        if (mediationExtras != null) {
            String keywordTargeting = mediationExtras.getString(MEDIATION_EXTRAS_SMART_KEYWORD_TARGETING_KEY);
            if (keywordTargeting != null) {
                targeting = keywordTargeting;
            }
        }

        return new SASAdPlacement(siteId, pageId, formatId, targeting);
    }



    @NonNull
    public static VersionInfo getVersionInfo() {
        return adapterVersionInfo;
    }

    @NonNull
    public static VersionInfo getSDKVersionInfo() {
        return sdkVersionInfo;
    }
}

File: com/smartadserver/android/library/mediation/SASGMACustomEventUtil.kt
package com.smartadserver.android.library.mediation

import android.content.Context
import android.os.Bundle
import com.google.android.gms.ads.mediation.*
import com.smartadserver.android.coresdk.util.SCSConfiguration
import com.smartadserver.android.library.model.SASAdPlacement
import com.smartadserver.android.library.util.SASConfiguration
import com.smartadserver.android.library.util.SASLibraryInfo

/**
 * Base class of all other SASCustomEvent adapters. Handle the SDK configuration.
 */
object SASGMACustomEventUtil {
    @JvmField
    val MEDIATION_EXTRAS_SMART_KEYWORD_TARGETING_KEY: String = "smart_keyword_targeting"

    // version info needed by google classes, init in a static block
    private var adapterVersionInfo: VersionInfo? = null
    private var sdkVersionInfo: VersionInfo? = null

    /**
     * Configure the Smart Display SDK if needed and
     *
     * @param context         The application context.
     * @param placementString The Smart placement String.
     * @param mediationExtras a Bundle containing mediation extra parameters as passed to the Google AdRequest in the application
     * @return a valid SASAdPlacement, or null if the SDK can not be configured or if the placement string is wrongly set.
     */
    fun configureSDKAndGetAdPlacement(context: Context,
                                      placementString: String,
                                      mediationExtras: Bundle?): SASAdPlacement? {
        var siteId: Int = -1
        var pageId: String = ""
        var formatId: Int = -1
        var targeting: String = ""

        // tokenize placement string and fill adPlacement;
        val ids = placementString.split("/")
        try {
            siteId = ids.get(0).trim().toInt()
            pageId = ids.get(1)
            formatId = ids.get(2).trim().toInt()
        } catch (e: Exception) {
            // invalid placement, return null
            return null
        }

        // configure the Smart Ad Server SDK if necessary
        if (!SASConfiguration.getSharedInstance().isConfigured()) {
            try {
                if (siteId >= 1) {
                    SASConfiguration.getSharedInstance().configure(context, siteId)
                    SASConfiguration.getSharedInstance().isPrimarySdk = false
                } else {
                    return null
                }
            } catch (e: SCSConfiguration.ConfigurationException) {
                e.printStackTrace()
                return null
            }
        }

        // extract custom targeting from mediation extras
        if (mediationExtras != null) {
            val keywordTargeting: String? = mediationExtras.getString(MEDIATION_EXTRAS_SMART_KEYWORD_TARGETING_KEY)
            if (keywordTargeting != null) {
                targeting = keywordTargeting
            }
        }
        return SASAdPlacement(siteId.toLong(), pageId, formatId.toLong(), targeting)
    }

    val versionInfo: VersionInfo
        get() {
            return (adapterVersionInfo)!!
        }
    val SDKVersionInfo: VersionInfo
        get() {
            return (sdkVersionInfo)!!
        }

    init {
        // adapters version info
        adapterVersionInfo = VersionInfo(1, 0, 0)

        //SDK version info
        val versionInfo = SASLibraryInfo.getSharedInstance().getVersion().split(".")

        // we expect 3 tokens, no more, no less
        val majorVersion = versionInfo.getOrNull(0)?.toIntOrNull() ?: 0
        val minorVersion = versionInfo.getOrNull(1)?.toIntOrNull() ?: 0
        val microVersion = versionInfo.getOrNull(1)?.toIntOrNull() ?: 0

        sdkVersionInfo = VersionInfo(majorVersion, minorVersion, microVersion)
    }
}